src/main/java/com/google/devtools/build/lib/rules/repository/RepoRecordedInput.java

```diff
@@ -106,121 +137,192 @@ public int compareTo(RepoRecordedInput o) {
-      PackageLookupValue pkgLookupValue = (PackageLookupValue) env.getValue(getSkyKey());
-      if (pkgLookupValue == null || !pkgLookupValue.packageExists()) {
-        return false;
+      Root root;
+      if (repoName.isMain()) {
+        root =
+            Root.fromPath(
+                PrecomputedValue.PATH_PACKAGE_LOCATOR.get(env).getPathEntries().get(0).asPath());
+      } else {
+        RepositoryDirectoryValue repoDirValue =
+            (RepositoryDirectoryValue) env.getValue(getSkyKey(env));
+        if (repoDirValue == null || !repoDirValue.repositoryExists()) {
+          return false;
+        }
+        root = Root.fromPath(repoDirValue.getPath());
       }
-      RootedPath rootedPath =
-          RootedPath.toRootedPath(pkgLookupValue.getRoot(), label.toPathFragment());
+      RootedPath rootedPath = RootedPath.toRootedPath(root, pathFragment);
       SkyKey fileKey = FileValue.key(rootedPath);
       try {
         FileValue fileValue = (FileValue) env.getValueOrThrow(fileKey, IOException.class);
         if (fileValue == null || !fileValue.isFile() || fileValue.isSpecialFile()) {
           return false;
         }
-        return oldValue.equals(RepositoryFunction.fileValueToMarkerValue(fileValue));
+        return oldValue.equals(fileValueToMarkerValue(fileValue));
       } catch (IOException e) {
         return false;
       }
     }
   }
 
   /**
-   * Represents a file input accessed during the repo fetch that is <i>not</i> addressable by a
-   * label. This most likely means that it's outside any known Bazel workspace.
+   * Represents a file input accessed during the repo fetch that is outside the current Bazel
+   * workspace. This file is addressed by its absolute path.
    */
-  public static final class AbsolutePathFile extends File {
-    final RootedPath path;
+  public static final class FileOutsideWorkspace extends File {
+    private final PathFragment path;
 
-    public AbsolutePathFile(RootedPath path) {
+    public FileOutsideWorkspace(PathFragment path) {
       this.path = path;
     }
 
     @Override
-    String toStringInternal() {
-      return path.asPath().getPathString();
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (!(o instanceof FileOutsideWorkspace)) return false;
+      FileOutsideWorkspace that = (FileOutsideWorkspace) o;
+      return Objects.equals(path, that.path);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(path);
+    }
+
+    @Override
+    public String toStringInternal() {
+      return path.getPathString();
     }
 
     @Override
-    public SkyKey getSkyKey() {
-      return FileValue.key(path);
+    public SkyKey getSkyKey(Environment env) throws InterruptedException {
+      return FileValue.key(
+          RootedPath.toRootedPath(
+              Root.absoluteRoot(
+                  PrecomputedValue.PATH_PACKAGE_LOCATOR.get(env).getOutputBase().getFileSystem()),
```

Comment:
nit: any reason why exactly this mechanism to get a `FileSystem`? There is already a `BlazeDirectories` instance in `RepositoryDelegatorFunction`, which would let you avoid an extra Skyframe dependency edge and `BlazeDirectories` is constant over the life time of the server anyways.

Reply:
RepoRecordedInput doesn't have access to a BlazeDirectories; we could add that to `RepoRecordedInput.parse` but that kind of pollutes the whole call chain. Getting it from Skyframe seemed like the easiest way (also, a Skyframe dependency on a precomputed value doesn't really cost anything IIUC)

Reply:
At risk of bikeshedding: you wouldn't need to add a method to `RepoRecordedInput.parse()`, only to `isUpToDate()` or `getSkyKey()`, both of which are called directly from `RepositoryFunction.verifyRecordedInputs()`, so you could add the `BlazeDirectories` as a member to `RepositoryFunction`. Would that work?

Reply:
ah yeah, that's true. It's definitely less ugly to pass `BlazeDirectories` to `getSkyKey()` than to pass `Environment` to it... done.